File: Days 041 - 050/Day 41/GetFlightItinerary.cpp
#include <iostream>
#include <string>
#include <utility>
#include <vector>

template <typename T>
std::string JoinVector(const std::vector<T>& values, const std::string& delimiter = "")
{
	std::string result;

	for (const auto& value : values)
	{
		result += value + delimiter;
	}

	return result;
}

std::vector<std::string> GetFlightItineraryHelper(const std::vector<std::pair<std::string, std::string>>& flights, const std::string& startingAirport, std::vector<std::string>& currentItinerary)
{
	if (flights.empty())
	{
		return currentItinerary.push_back(startingAirport), currentItinerary;
	}

	std::vector<std::string> itinerary;

	for (unsigned int i = 0; i < flights.size(); i++)
	{
		auto[startCity, endCity] = flights[i];

		if (startingAirport == startCity)
		{
			currentItinerary.push_back(startCity);

			std::vector<std::pair<std::string, std::string>> continuedFlights;

			for (unsigned int j = 0; j < flights.size(); j++)
			{
				if (j != i)
				{
					continuedFlights.push_back(flights[j]);
				}
			}

			std::vector<std::string> continuedItinerary = GetFlightItineraryHelper(continuedFlights, endCity, currentItinerary);

			if (!continuedItinerary.empty())
			{
				if (itinerary.empty() || (JoinVector(continuedItinerary) < JoinVector(itinerary)))
				{
					itinerary = continuedItinerary;
				}
			}
		}
	}

	return itinerary;
}

std::vector<std::string> GetFlightItinerary(const std::vector<std::pair<std::string, std::string>>& flights, const std::string& startingAirport)
{
	std::vector<std::string> itinerary;

	return GetFlightItineraryHelper(flights, startingAirport, itinerary);
}

template <typename T>
void PrintVector(const std::vector<T>& values)
{
	if (values.empty())
	{
		std::cout << "Vector is empty...\n";

		return;
	}

	for (const auto& value : values)
	{
		std::cout << value << " ";
	}

	std::cout << "\n";
}

int main(int argc, const char* argv[])
{
	std::vector<std::pair<std::string, std::string>> flights {
		{ "SFO", "HKO" },
		{ "YYZ", "SFO" },
		{ "YUL", "YYZ"},
		{ "HKO", "ORD" }
	};

	PrintVector(GetFlightItinerary(flights, "YUL"));

	flights = {
		{ "SFO", "COM" },
		{ "COM", "YYZ" }
	};

	PrintVector(GetFlightItinerary(flights, "COM"));

	flights = {
		{ "A", "B" },
		{ "A", "C" },
		{ "B", "C" },
		{ "C", "A" }
	};

	PrintVector(GetFlightItinerary(flights, "A"));

	std::cin.get();

	return 0;
}

File: Days 041 - 050/Day 41/GetFlightItinerary.cs
using System;
using System.Collections.Generic;

namespace DailyCodingProblem
{
	internal class Day41
	{
		private static int Main(string[] args)
		{
			List<(string, string)> flights = new List<(string, string)>()
			{
				("SFO", "HKO"),
				("YYZ", "SFO"),
				("YUL", "YYZ"),
				("HKO", "ORD")
			};

			PrintList(GetFlightItinerary(flights, "YUL"));

			flights = new List<(string, string)>()
			{
				("SFO", "COM"),
				("COM", "YYZ")
			};

			PrintList(GetFlightItinerary(flights, "COM"));

			Console.ReadLine();

			return 0;
		}

		private static List<string> GetFlightItinerary(List<(string, string)> flights, string startingAirport)
		{
			List<string> itinerary = new List<string>();

			return GetFlightItineraryHelper(flights, startingAirport, itinerary);
		}

		private static List<string> GetFlightItineraryHelper(List<(string, string)> flights, string startingAirport, List<string> currentItinerary)
		{
			if (flights.Count == 0)
			{
				currentItinerary.Add(startingAirport);
				return currentItinerary;
			}

			List<string> itinerary = new List<string>();

			for (int i = 0; i < flights.Count; i++)
			{
				(string startCity, string endCity) = flights[i];

				if (startingAirport == startCity)
				{
					currentItinerary.Add(startCity);
					List<(string, string)> continuedFlights = new List<(string, string)>();

					for (int j = 0; j < flights.Count; j++)
					{
						if (j != i)
						{
							continuedFlights.Add(flights[j]);
						}
					}

					List<string> continuedItinerary = GetFlightItineraryHelper(continuedFlights, endCity, currentItinerary);

					if (continuedItinerary.Count > 0)
					{
						if (itinerary.Count == 0 || string.Compare(JoinListAsString(continuedItinerary), JoinListAsString(itinerary)) < 0)
						{
							itinerary = continuedItinerary;
						}
					}
				}
			}

			return itinerary;
		}

		private static string JoinListAsString<T>(List<T> values, string delimiter = "") => string.Join(delimiter, values.ToArray());

		private static void PrintList<T>(List<T> values)
		{
			if (values.Count == 0)
			{
				Console.WriteLine("List is empty...");

				return;
			}

			foreach (T value in values)
			{
				Console.Write($"{value} ");
			}

			Console.WriteLine();
		}
	}
}